src/ir/ExpressionManipulator.cpp

```diff
@@ -23,306 +23,90 @@ namespace ExpressionManipulator {
-      return builder.makeI31Get(copy(curr->i31), curr->signed_);
-    }
-    Expression* visitRefTest(RefTest* curr) {
-      WASM_UNREACHABLE("TODO (gc): ref.test");
-    }
-    Expression* visitRefCast(RefCast* curr) {
-      WASM_UNREACHABLE("TODO (gc): ref.cast");
-    }
-    Expression* visitBrOnCast(BrOnCast* curr) {
-      WASM_UNREACHABLE("TODO (gc): br_on_cast");
-    }
-    Expression* visitRttCanon(RttCanon* curr) {
-      WASM_UNREACHABLE("TODO (gc): rtt.canon");
-    }
-    Expression* visitRttSub(RttSub* curr) {
-      WASM_UNREACHABLE("TODO (gc): rtt.sub");
-    }
-    Expression* visitStructNew(StructNew* curr) {
-      WASM_UNREACHABLE("TODO (gc): struct.new");
-    }
-    Expression* visitStructGet(StructGet* curr) {
-      WASM_UNREACHABLE("TODO (gc): struct.get");
-    }
-    Expression* visitStructSet(StructSet* curr) {
-      WASM_UNREACHABLE("TODO (gc): struct.set");
-    }
-    Expression* visitArrayNew(ArrayNew* curr) {
-      WASM_UNREACHABLE("TODO (gc): array.new");
-    }
-    Expression* visitArrayGet(ArrayGet* curr) {
-      WASM_UNREACHABLE("TODO (gc): array.get");
-    }
-    Expression* visitArraySet(ArraySet* curr) {
-      WASM_UNREACHABLE("TODO (gc): array.set");
-    }
-    Expression* visitArrayLen(ArrayLen* curr) {
-      WASM_UNREACHABLE("TODO (gc): array.len");
-    }
-  };
+#define DELEGATE_FIELD_CHILD_LIST(id, name)                                    \
+  castCopy->name.resize(castOriginal->name.size());                            \
+  for (Index i = 0; i < castOriginal->name.size(); i++) {                      \
+    tasks.push_back({castOriginal->name[i], &castCopy->name[i]});              \
+  }
+
+#define DELEGATE_FIELD_INT(id, name) castCopy->name = castOriginal->name;
+
+#define DELEGATE_FIELD_INT_ARRAY(id, name)                                     \
+  assert(castCopy->name.size() == castOriginal->name.size());                  \
+  for (Index i = 0; i < castOriginal->name.size(); i++) {                      \
+    castCopy->name[i] = castOriginal->name[i];                                 \
+  }
+
+#define DELEGATE_FIELD_LITERAL(id, name) castCopy->name = castOriginal->name;
 
-  Copier copier(wasm, custom);
-  return copier.copy(original);
+#define DELEGATE_FIELD_NAME(id, name) castCopy->name = castOriginal->name;
+
+#define DELEGATE_FIELD_SCOPE_NAME(id, name) castCopy->name = castOriginal->name;
+
+#define DELEGATE_FIELD_SCOPE_NAME_LIST(id, name)                               \
+  castCopy->name.resize(castOriginal->name.size());                            \
+  for (Index i = 0; i < castOriginal->name.size(); i++) {                      \
+    castCopy->name[i] = castOriginal->name[i];                                 \
+  }
+
+#define DELEGATE_FIELD_SIGNATURE(id, name) castCopy->name = castOriginal->name;
+
+#define DELEGATE_FIELD_TYPE(id, name) castCopy->name = castOriginal->name;
+
+#define DELEGATE_FIELD_ADDRESS(id, name) castCopy->name = castOriginal->name;
```

Comment:
It looks we can reduce the number of `define`s here by doing something like
```cpp
#define COPY_FIELD(id, name) castCopy->name = castOriginal->name;

#define COPY_FIELD_LIST(id, name)                                              \
  for (Index i = 0; i < castOriginal->name.size(); i++) {                      \
    castCopy->name[i] = castOriginal->name[i];                                 \
  }

#define DELEGATE_FIELD_INT COPY_FIELD
#define DELEGATE_FIELD_INT_ARRAY(id, name)                                     \
  assert(castCopy->name.size() == castOriginal->name.size());                  \
  COPY_FIELD_LIST(id, name)
#define DELEGATE_FIELD_LITERAL COPY_FIELD
#define DELEGATE_FIELD_NAME COPY_FIELD
#define DELEGATE_FIELD_SCOPE_NAME COPY_FIELD
#define DELEGATE_FIELD_SCOPE_NAME_LIST(id, name)                               \
  castCopy->name.resize(castOriginal->name.size());                            \
  COPY_FIELD_LIST(id, name)
#define DELEGATE_FIELD_SIGNATURE COPY_FIELD
#define DELEGATE_FIELD_TYPE COPY_FIELD
#define DELEGATE_FIELD_ADDRESS COPY_FIELD
```

Reply:
Good idea, thanks!

I also renamed the variable-sized one "vector", to match the fixed-size one which is "array", which is hopefully more like C++ and less confusing.